ondewo_bpi_qa/bpi_qa_server.py

```diff
@@ -79,18 +89,84 @@ def check_session_id(self, request: DetectIntentRequest) -> None:
                 "loop": asyncio.new_event_loop(),
                 "timestamp": time.time(),
             }
-            logger_console.warning(
+            logger_console.debug(
                 f"New session in bpi: {request.session}. {len(self.loops)} sessions currently stored."
             )
 
-    @Timer(log_arguments=False)
-    def handle_async(self, request: DetectIntentRequest,) -> Tuple[DetectIntentResponse, str]:
+    @Timer(log_arguments=False, logger=logger_console.debug)
+    def handle_context_injection_for_qa(self, request: DetectIntentRequest, ) -> DetectIntentRequest:
+        """
+        Note: to enable Q&A to leverage context injection, we made the context injection happen BEFORE
+            detect intent is called. This way we can guarantee on our async calls that both Q&A and CAI
+            have the same contexts set.
+
+        Note 2: By the nature of it, this function contains a SIDE-EFFECT! It modifies the
+            DetectIntentRequest by removing the context "c-qa-url-filter" if found, this context
+            is managed manually instead.
+        """
+        parent: str = ContextHelper.get_agent_path_from_path(request.session)
+        unaffected_contexts: List[Context] = []
+
+        if not request.query_params:
+            return request
+
+        try:
+            qa_url_context: Optional[Context] = self.client.services.contexts.get_context(
+                GetContextRequest(name=f'{request.session}/contexts/{QA_URL_FILTER_CONTEXT_NAME}')
+            )
+        except Exception:
+            qa_url_context = None
+
+        for context in request.query_params.contexts:
+
+            if QA_URL_FILTER_CONTEXT_NAME not in context.name:
+                unaffected_contexts.append(context)
+                continue
+
+            context_to_inject: Context = Context(
+                name=context.name,
```

Comment:
Can't you just do
context_to_inject: Context = Context()
context_to_inject.CopyFrom(context)

 So it's kinda simpler than writing everything out